web3_proxy/src/frontend/users/payment.rs

```diff
@@ -413,60 +416,51 @@ pub async fn user_balance_post(
-                    // Make the recipient premium "Effectively Unlimited"
-                    active_recipient.user_tier_id = sea_orm::Set(premium_user_tier.id);
-                    active_recipient.save(&txn).await?;
-                }
-
-                debug!("New user balance model is: {:?}", active_user_balance);
-                active_user_balance.save(&txn).await?;
-                // txn.commit().await?;
-                // user_balance
-            }
-            None => {
-                // Create the entry with the respective balance
-                let active_user_balance = balance::ActiveModel {
-                    available_balance: sea_orm::ActiveValue::Set(amount),
-                    user_id: sea_orm::ActiveValue::Set(recipient.id),
-                    ..Default::default()
-                };
-
-                if amount >= Decimal::new(10, 0) {
-                    // Also make the user premium at this point ...
-                    let mut active_recipient = recipient.clone().into_active_model();
-                    // Make the recipient premium "Effectively Unlimited"
-                    active_recipient.user_tier_id = sea_orm::Set(premium_user_tier.id);
-                    active_recipient.save(&txn).await?;
-                }
-
-                info!("New user balance model is: {:?}", active_user_balance);
-                active_user_balance.save(&txn).await?;
-                // txn.commit().await?;
-                // user_balance // .try_into_model().unwrap()
-            }
+        let balance_entry = balance::ActiveModel {
+            id: sea_orm::NotSet,
+            available_balance: sea_orm::Set(amount),
+            used_balance: sea_orm::Set(Decimal::new(0, 0)),
+            user_id: sea_orm::Set(recipient.id),
         };
+
+        // TODO: Do an insert on conflict update ...
+        let balance_entry_id = balance::Entity::insert(balance_entry)
+            .on_conflict(
+                OnConflict::new()
+                    .values([(
+                        balance::Column::AvailableBalance,
+                        Expr::col(balance::Column::AvailableBalance).add(amount),
+                    )])
+                    .to_owned(),
+            )
+            .exec(&txn)
+            .await?
+            .last_insert_id;
+
+        let new_balance_entry = balance::Entity::find()
+            .filter(balance::Column::Id.eq(balance_entry_id))
+            .one(&txn)
+            .await?
+            .context("Could not find the newly inserted balance_entry, something went wrong!")?;
+
+        // Then also make the user premium if the payment is above 10$
+        if new_balance_entry.available_balance >= Decimal::new(10, 0) {
+            let mut active_recipient = recipient.clone().into_active_model();
+            // Make the recipient premium "Effectively Unlimited"
+            active_recipient.user_tier_id = sea_orm::Set(premium_user_tier.id);
```

Comment:
Don't actually change the row database. The row should always be their actual tier. 

Instead, the code that fetches the rate limits should check the balance. If the balance is low, use the out_of_funds_tier_id for the user. 

Changing it in the actual database will cause problems when we have many tiers

Reply:
Again, we still do not need this here. We don't want balance changes changing the column ever. Delete this. 

The way it should work, is that when a user registers, they are given the default premium role. As they won't a balance yet, the code will give them the limits of the out of funds tier that matches this role. As soon as they do have a balance, they code will see that, and they will get the limits of the premium role.

Reply:
sorry for this! for some reason that's an artefcat that wasnt removed during the last merge ... will remove this once and for all now!

Reply:
oh wait nevermind, this was a confusion from my side. logic makes sense now (that premium is default basically)

Reply:
as a small heads up though, rn the code does not make the users premium by default, should that be also added?